examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/solver_gmres.h>
+
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/tria_boundary_lib.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/grid_refinement.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_renumbering.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/error_estimator.h>
+
+#include <deal.II/lac/sparse_direct.h>
+
+#include <deal.II/lac/sparse_ilu.h>
+#include <deal.II/grid/grid_out.h>
+
+// We need to include this class for all of the timings between ILU and Multigrid
+#include <deal.II/base/timer.h>
+
+// This includes the files necessary for us to use geometric Multigrid
+#include <deal.II/multigrid/multigrid.h>
+#include <deal.II/multigrid/mg_transfer.h>
+#include <deal.II/multigrid/mg_tools.h>
+#include <deal.II/multigrid/mg_coarse.h>
+#include <deal.II/multigrid/mg_smoother.h>
+#include <deal.II/multigrid/mg_matrix.h>
+
+#include <fstream>
+#include <sstream>
+
+namespace Step55
+{
+  using namespace dealii;
+
+  // In order to make it easy to switch between the different solvers that are being
+  // used in Step-55, an enum was created that can be passed as an argument to the
+  // constructor of the main class.
+  struct SolverType
+  {
+    enum type {FGMRES_ILU, FGMRES_GMG, UMFPACK};
+  };
+
+  // @sect3{Functions for Solution and Righthand side}
+  //
+  // The class Solution is used to define the boundary conditions and to
+  // compute errors of the numerical solution. Note that we need to define
+  // the values and gradients in order to compute L2 and H1 errors. Here
+  // we decided to separate the implementations for 2d and 3d using
+  // template specialization.  We do this to make it easier for us to debug
+  // as well as its aesthetic value.
+  template <int dim>
+  class Solution : public Function<dim>
+  {
+  public:
+    Solution () : Function<dim>(dim+1) {}
+    virtual double value (const Point<dim> &p,
+                          const unsigned int component) const;
+    virtual Tensor<1,dim> gradient (const Point<dim> &p,
+                                    const unsigned int component = 0) const;
+  };
+
+  template <>
+  double
+  Solution<2>::value (const Point<2> &p,
+                      const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+
+    if (component == 0)
+      return sin (PI * x);
+    if (component == 1)
+      return - PI * y * cos(PI * x);
+    if (component == 2)
+      return sin (PI * x) * cos (PI * y);
```

Comment:
Added into the description of solution above the class to avoiding repeating it many times.